### PR TITLE
Fix thread safety of the translog indexer, remove shared state

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Follow up of https://github.com/crate/crate/commit/10dee44.

Applying [suggestion](https://github.com/crate/crate/pull/17403#issuecomment-2651408103) from @seut , thanks!

### CODE REVIEW

**Comment by seut** [comment]
Given that I authored this code, someone else should review it ;)

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java [hunk 1]
```diff
@@ -85,19 +84,29 @@ public TranslogIndexer(DocTableInfo table, Version shardCreatedVersion) {
...
+            reference -> new IndexInput(reference.column())
         );
         ignoreUnknownColumns = table.columnPolicy() != ColumnPolicy.STRICT;
         this.shardCreatedVersion = shardCreatedVersion;
     }
 
 
-    private record IndexInput(ColumnIdent columnIdent,
-                              Supplier<Map<String, Object>> sourceSupplier) implements Input<Object> {
+    private static class IndexInput implements Input<Object> {
```
Comment (romseygeek) [inline server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: Does this actually need to implement `Input<>`?  If we can avoid that then we don't need the unimplemented method, which feels a bit clunky.

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java [hunk 2]
```diff
@@ -138,7 +147,27 @@ private IndexDocumentBuilder populateLuceneFields(BytesReference source) throws
             entry.setValue(castValue);
         }
 
-        addIndexColumns(indexColumns, docBuilder);
+        for (var indexColumn : indexColumns) {
```
Comment (romseygeek) [inline server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:150]: I'd pull this into a separate method, as otherwise `populateLuceneFields` gets a bit too big to easily read.

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java [hunk 3]
```diff
@@ -138,7 +147,27 @@ private IndexDocumentBuilder populateLuceneFields(BytesReference source) throws
...
+            String fqn = indexColumn.reference().storageIdent();
+            for (var input : indexColumn.inputs()) {
+                Object value = input.value(docMap);
+                if (value == null) {
+                    continue;
+                }
```
Comment (romseygeek) [inline server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:156]: Can we add some tests that cover null values via `assertTranslogParses`?

**Review by romseygeek — CHANGES_REQUESTED** [review]
Nearly there, I left a couple of comments.

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: We won't be able to re-use Indexer's `buildIndexColumns` which is probably not a big deal (but would have to duplicate the logic). Maybe I should add the javadoc to the class explaining why we made it implement `Input` or better go with custom logic?

Comment (romseygeek) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: Do we need the `<I extends Input<?>>` restriction in `buildIndexColumns`?

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:156]: Thanks, pushed a null for array, top-column and sub-column, have a failure on sub-column that fails on assertTranslogParses but could be a test setup issue, checking

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:156]: Yep, it's unrelated to FT, simple sub-column with NULL fails on `assertTranslogParses` on master

File: server/src/main/java/io/crate/execution/dml/TranslogIndexer.java (hunk 1 above)
Comment (seut) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: If we want to reuse `Indexer.buildIndexColumns`, yes.
As already said, I'm not sure if its worth doing.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: No, at the end we have to use Input as essentially it all boils down to the Indexer's logic operating with inputs, ie we use expressions derived from `ctxForRefs::add `to fetch the value.

Suggestions:
1. Don't use Input at all in the TranslogIndexer and get value from the docMap directly as in https://github.com/crate/crate/pull/17403/commits/f48b17cbab0a78104541dd0eed8b4459f74eace8#diff-d08b3d5fd51f3ae7aeb3cc7d346da46d6b9aeefc68d7da5a0cbbdaf6e719fd5cR126-R132

but then ned to introduce new Map to precompute fullPath, here it was nicely encapsulated in the Input + allowed to re-use Indexer's logic (maybe not important, not sure).
2. Let it implement Input , ie leave as is, maybe add a code comment explaining why it implements Input
3. Something else?

Comment (romseygeek) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: Right, Indexer needs to use an `Input<?>`, but `IndexColumn` itself is just a parametrized container.  So Indexer can use `IndexColumn<Input<?>>` and TranslogIndexer can use `IndexColumn<IndexInput>`; the final argument passed to `buildIndexColumn` determines the parameter type.

Comment (romseygeek) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java:94]: In other words, I think you can replace
```
    public static <I extends Input<?>> List<IndexColumn<I>> buildIndexColumns( ...
```
with
```
public static <I> List<IndexColumn<I>> buildIndexColumns( ...
```
and everything will still work in Indexer, but IndexInput doesn't need to implement Input anymore.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/dml/TranslogIndexer.java]: Right, thanks, pushed an update

**Comment by BaurzhanSakhariev** [comment]
Unrelated https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/19287/testReport/junit/io.crate.lucene.index/TestShuffleForcedMergePolicyTests/testDiagnostics/

retest this please
